File: scripts/threshold-workout.js
// Define the threshold power (in watts)
const thresholdPower = 250;

// Define the duration of the intervals (in seconds)
const intervalDuration = 300;

// Define the number of intervals
const numIntervals = 5;

// Define the rest period between intervals (in seconds)
const restDuration = 60;

// Define the warm-up period (in seconds)
const warmupDuration = 600;

// Define the cool-down period (in seconds)
const cooldownDuration = 600;

// Define the total workout duration (in seconds)
const totalDuration = (numIntervals * intervalDuration) + ((numIntervals - 1) * restDuration) + warmupDuration + cooldownDuration;

// Define the power output for each interval (as a percentage of threshold power)
const intervalPower = 90;

// Define the power output for the rest period (as a percentage of threshold power)
const restPower = 50;

// Define the power output for the warm-up and cool-down periods (as a percentage of threshold power)
const warmupCoolDownPower = 60;

// Define a function to calculate the power output for each time interval
function calculatePowerOutput(currentTime) {
  let powerOutput;
  
  // Warm-up period
  if (currentTime < warmupDuration) {
    powerOutput = warmupCoolDownPower;
  }
  // Cool-down period
  else if (currentTime >= totalDuration - cooldownDuration) {
    powerOutput = warmupCoolDownPower;
  }
  // Intervals
  else {
    const intervalIndex = Math.floor((currentTime - warmupDuration) / (intervalDuration + restDuration));
    
    if (intervalIndex % 2 === 0) {
      powerOutput = intervalPower;
    } else {
      powerOutput = restPower;
    }
  }
  
  // Convert power output from a percentage of threshold power to watts
  powerOutput = thresholdPower * (powerOutput / 100);
  
  return powerOutput;
}

// Define a function to simulate the workout
function simulateWorkout() {
  // Set the initial time to zero
  let currentTime = 0;
  
  // Loop until the total duration of the workout is reached
  while (currentTime < totalDuration) {
    // Calculate the power output for the current time
    const powerOutput = calculatePowerOutput(currentTime);
    
    // Send the power output to the smart bike
    sendPowerOutputToBike(powerOutput);
    
    // Wait for one second before updating the current time
    await sleep(1000);
    
    // Update the current time
    currentTime += 1;
  }
  
  // Turn off the smart bike
  turnOffBike();
}

// Define a function to send the power output to the smart bike
function sendPowerOutputToBike(powerOutput) {
  // Send the power output to the smart bike
  // TODO: Implement this function
}

// Define a function to turn off the smart bike
function turnOffBike() {
  // Turn off the smart bike
  // TODO: Implement this function
}

// Define a function to pause execution for a given number of milliseconds
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// Call the simulateWorkout function to start the workout
simulateWorkout();

File: Drivers/Proof_of_Concepts/ThresholdWorkout.cs
using System;
using System.Threading.Tasks;

class ThresholdWorkout {
    // Define the constants for the workout
    const int IntervalDuration = 20; // in seconds
    static int NumIntervals ;
    const int RestDuration = 15; // in seconds
    const int WarmupDuration = 20; // in seconds
    const int CooldownDuration = 10; // in seconds

    // Define the power output percentages for each part of the workout
    static int ThresholdPower;
    const double IntervalPowerPercentage = 0.90; // 90% of the threshold power
    const double RestPowerPercentage = 0.50; // 50% of the threshold power
    const double WarmupPowerPercentage = 0.30; // 30% of the threshold power
    const double CooldownPowerPercentage = 0.30; // 30% of the threshold power

    // Define a function to calculate the power output for a given time during the workout
    static int CalculatePowerOutput(int time) {
        if (time < WarmupDuration) {
            return (int)(ThresholdPower * WarmupPowerPercentage);
        } else if (time < WarmupDuration + NumIntervals * (IntervalDuration + RestDuration)) {
            int intervalIndex = (time - WarmupDuration) / (IntervalDuration + RestDuration);
            int intervalStartTime = WarmupDuration + intervalIndex * (IntervalDuration + RestDuration);
            int intervalElapsedTime = time - intervalStartTime;
            if (intervalElapsedTime < IntervalDuration) {
                return (int)(ThresholdPower * IntervalPowerPercentage);
            } else {
                return (int)(ThresholdPower * RestPowerPercentage);
            }
        } else {
            return (int)(ThresholdPower * CooldownPowerPercentage);
        }
    }

    // Define a function to simulate the workout
    static async Task SimulateWorkout() {
        Console.WriteLine("Starting workout...");
        await Task.Delay(2000); // Wait for 2 seconds before starting to simulate the workout

        int currentTime = 0;
        int TotalDuration = (NumIntervals * IntervalDuration) + (NumIntervals - 1) * RestDuration + WarmupDuration + CooldownDuration;

        while (currentTime < TotalDuration) {
            int powerOutput = CalculatePowerOutput(currentTime);
            Console.WriteLine($"Power output for time {currentTime} is {powerOutput}");

            await Task.Delay(1000); // Wait for 1 second
            currentTime += 1;
        }

        Console.WriteLine("Workout complete!");
       
    }

    // Define a function to turn off the smart bike
    static void TurnOffBike() {
        Console.WriteLine("Smart bike turned off.");
    }

    // Main function
    static async Task Main() {
        // Take input from user about the threshold power
        Console.Write("Enter the threshold power (in watts): ");
        int.TryParse(Console.ReadLine(), out ThresholdPower);
        // Take input from user about number of intervals
        Console.Write("Please enter the number of intervals for the workout");
        int.TryParse(Console.ReadLine(), out NumIntervals );

        // Simulate the workout
        await SimulateWorkout();

        // Turn off the smart bike
        TurnOffBike();
    }
}
